folder/folderutil.go

```diff
@@ -145,13 +145,30 @@ func agnosticSplit(path string) (parts []string) {
 	return
 }
 
-// HomeDirectory returns the home directory or defaultDirectory in case of error
+// HomeDirOrDefault tries to obtain the user's home directory and
+// returns the default if it cannot be obtained.
 func HomeDirOrDefault(defaultDirectory string) string {
-	usr, err := user.Current()
+	if user, err := user.Current(); err == nil && isWritable(user.HomeDir) {
+		return user.HomeDir
+	}
+	if homeDir, err := os.UserHomeDir(); err == nil && isWritable(homeDir) {
+		return homeDir
+	}
+	return defaultDirectory
+}
+
+// isWritable checks if a path is writable.
+func isWritable(path string) bool {
+	if !fileutil.FolderExists(path) {
+		return false
+	}
+
+	tmpfile, err := os.CreateTemp(path, "test")
```

Comment:
Probably in unix we can use `golang.org/x/sys/unix.Access(...)` but I agree that whether a bit invasive, this method always work. Maybe we can add within the comment that it's preferred to declare a global variable in the module using it and initialize it once as the function involves a file creation check. What do you think?

Reply:
That sounds reasonable. I've made some modifications. Could you take another look at it?